gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorage.java

```diff
@@ -200,6 +206,46 @@ void copy(
       List<String> dstObjectNames)
       throws IOException;
 
+  /**
+   * Copies metadata of the given objects. After the copy is successfully complete, each object blob
+   * is reachable by two different names. Copying between two different locations or between two
+   * different storage classes is not allowed.
+   *
+   * @param sourceToDestinationObjectsMap map of destination objects to be copied, keyed by source
+   * @throws java.io.FileNotFoundException if the source object or the destination bucket does not
+   *     exist
+   * @throws IOException in all other error cases
+   */
+  default void copy(Map<StorageResourceId, StorageResourceId> sourceToDestinationObjectsMap)
+      throws IOException {
+    checkArgument(
+        sourceToDestinationObjectsMap != null, "sourceToDestinationObjectsMap must not be null");
+    if (sourceToDestinationObjectsMap.isEmpty()) {
+      return;
+    }
+
+    List<String> srcObjectNames = new ArrayList<>();
+    List<String> dstObjectNames = new ArrayList<>();
```

Comment:
nit: initialize list size base on map size to optimize memory allocation a little